common/datavalues2/src/columns/builder.rs

```diff
@@ -14,27 +12,77 @@ use common_arrow::arrow::bitmap::Bitmap;
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-pub trait ColumnBuilder<N, C> {
-    fn append_value(&mut self, val: N);
-    fn append_null(&mut self);
-    fn append_option(&mut self, opt_val: Option<N>) {
-        match opt_val {
-            Some(v) => self.append_value(v),
-            None => self.append_null(),
+use std::sync::Arc;
+
+/// A builder for a column.
+use common_arrow::arrow::bitmap::MutableBitmap;
+
+use crate::prelude::MutableColumn;
+use crate::ColumnRef;
+use crate::ConstColumn;
+use crate::NullableColumn;
+use crate::ScalarType;
+
+pub struct ColumnBuilder<T: ScalarType> {
+    builder: T::MutableColumnType,
+    validity: MutableBitmap,
+}
+
+impl<T> ColumnBuilder<T>
+where T: ScalarType + Default
+{
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            builder: <T::MutableColumnType>::with_capacity(capacity),
+            validity: MutableBitmap::with_capacity(capacity),
         }
     }
-    fn finish(&mut self) -> C;
-}
 
-pub trait NewColumn<N> {
-    /// create non-nullable column by values
-    fn new_from_slice<P: AsRef<[N]>>(v: P) -> Self;
-    fn new_from_iter(it: impl Iterator<Item = N>) -> Self;
-}
+    #[inline]
+    pub fn append_null(&mut self) {
+        self.builder.append_default();
+        self.validity.push(false);
+    }
+
+    #[inline]
+    pub fn append_value(&mut self, value: T) {
+        self.builder.append(value);
+        self.validity.push(true);
+    }
 
-pub trait NewNullableColumn<N> {
-    /// create nullable column by option values
-    fn new_from_opt_slice(opt_v: &[Option<N>]) -> Self;
-    fn new_from_opt_iter(it: impl Iterator<Item = Option<N>>) -> Self;
-    fn new_from_iter_validity(it: impl Iterator<Item = N>, validity: Option<Bitmap>) -> Self;
+    #[inline]
+    pub fn append(&mut self, value: T, valid: bool) {
+        self.builder.append(value);
+        self.validity.push(valid);
+    }
+
+    #[inline]
+    pub fn build(&mut self, length: usize, nullable: bool) -> ColumnRef {
+        let column = self.build_nonull(length);
+        if nullable {
+            return Arc::new(NullableColumn::new(
+                column,
+                std::mem::take(&mut self.validity).into(),
+            ));
+        }
+        column
+    }
+
+    fn build_nonull(&mut self, length: usize) -> ColumnRef {
+        let col = self.builder.as_column();
+        if length != self.len() {
+            return Arc::new(ConstColumn::new(col, length));
```

Comment:
to make it a ConstColumn, maybe we will need assert `col.len() == 1` ?

Reply:
Right. I already fixed it, will create a fix in another pr.

```
fn build_nonull(&mut self, length: usize) -> ColumnRef {
        let col = self.builder.as_column();
        if length != self.len() && self.len() == 1 {
            return Arc::new(ConstColumn::new(col, length));
        }
        col
    }
```